src/trading/utils.ts

```diff
@@ -0,0 +1,82 @@
+import { AccountAddress, LimitTradeParameters, PrivateKey, TradeParameters } from './types'
+import { QuoteAmountsAndCosts } from '../order-book'
+import { ETH_ADDRESS } from '../common'
+import { ethers, Signer } from 'ethers'
+import { type ExternalProvider, Web3Provider } from '@ethersproject/providers'
+
+export function swapParamsToLimitOrderParams(
+  params: TradeParameters,
+  quoteId: number,
+  amounts: QuoteAmountsAndCosts
+): LimitTradeParameters {
+  return {
+    ...params,
+    sellAmount: amounts.afterSlippage.sellAmount.toString(),
+    buyAmount: amounts.afterSlippage.buyAmount.toString(),
+    quoteId,
+  }
+}
+
+export function getIsEthFlowOrder(params: { sellToken: string }): boolean {
+  return params.sellToken.toLowerCase() === ETH_ADDRESS.toLowerCase()
+}
+
+export function getSigner(signer: Signer | ExternalProvider | PrivateKey): Signer {
+  if (typeof signer === 'string') return new ethers.Wallet(signer)
+
+  if ('request' in signer || 'send' in signer) {
+    const provider = new Web3Provider(signer)
+
+    return provider.getSigner()
+  }
+
+  return signer as Signer
+}
+
+export function isAccountAddress(address: unknown): address is AccountAddress {
+  return typeof address === 'string' && /^0x[0-9a-fA-F]{40}$/.test(address)
+}
+
+/**
+ * Returns the gas value plus a margin for unexpected or variable gas costs (20%)
+ * @param value the gas value to pad
+ */
+export function calculateGasMargin(value: bigint): bigint {
+  return value + (value * BigInt(20)) / BigInt(100)
+}
+
+export function mapQuoteAmountsAndCosts<T, R>(
+  value: QuoteAmountsAndCosts<T>,
+  mapper: (value: T) => R
+): QuoteAmountsAndCosts<R> {
+  const {
+    costs: { networkFee, partnerFee },
+  } = value
+
+  function serializeAmounts(value: { sellAmount: T; buyAmount: T }): { sellAmount: R; buyAmount: R } {
+    return {
+      sellAmount: mapper(value.sellAmount),
+      buyAmount: mapper(value.buyAmount),
+    }
+  }
```

Comment:
Does this fn declaration need to be inside the parent fn?

Reply:
It depends on `mapper`. I could pass it as a parameter, but it doesn't look reasonable because it will require more code